### DISCUSSION TITLE
Two questions about developing gap8 with CPX

### DISCUSSION THREAD

**Discussion by RavenLite** [bitcraze/discussions#642] [Q&A]
Hi everyone,

These days, we are developing some applications that need a large amount of data to transfer (transfer point cloud from multi-ranger deck --> stm32 --> gap8). So, we are using CPX to do it. But while developing, I have two questions about CPX.
1. Why the CPX routing is stm32-->esp32-->gap8, not stm32-->gap8 directly? 
The esp32 module sometimes will not work correctly, so, I'm curious why CPX didn't been designed to route the data directly to gap8. I remember there are 2 UART that could be used separately by gap8 or esp32.

2. How to send/receive data automatically?
In our program, gap8 not only receives point clouds from stm32 but also sends results to stm32. So, gap8 cannot be in the listening status all the time, and stm32 also cannot be in the sending status all the time. I think the API provided by CPX can't help us to handle the switch now. Maybe we need to develop a controller to switch the listening/sending status? It's easy to switch the status of gap8 or stm32. But how to make them switch at the same time (means not listening at the same time or not sending at the same time)?

Thanks!
Raven

**Comment by hmllr** [comment]
Hi Raven,
1. Maybe @ataffanel knows it in more detail - but I would say it just simplifies things. Having two routes would be more complicated to handle, and the route over the ESP is anyway needed (as the ESP can update the gap8 firmware), plus it has the higher bandwidth between gap8 and ESP, which is important for image streaming, as this is SPI and not UART as the other links. You can still use the direct UART link though.
2. are you talking about sending the data over UART directly or over the ESP now? For only UART I don't think this should be a problem, you can use the DMA (or on the gap8 I think it is actually called uDMA, but that's a detail) to receive/send, such that it is not blocking (https://greenwaves-technologies.com/manuals/BUILD/PMSIS_API/html/group__UART.html#gada5035f9fd4b858f8326c8b30d254e56). For CPX I don't think it should be an issue either - the ESP-GAP8 communication has a handshake with an GPIO to avoid such problems and I think the STM32-ESP is not blocking. And even if you would want to use blocking functions - as you are sending point clouds in one direction and results in the other, meaning the packet order is well defined, it seems to be possible to figure out when to switch mode.  

Hope I understood the second question right and this helps,
 Hanna
